### ISSUE TITLE
[OPEN UI] Skeleton aria role proposal

### ISSUE THREAD

**Issue by gregwhitworth** [w3c/aria#1317]
There is currently a discussion occurring in Open UI and as promised if there are accessibility discussions I would ensure to loop you all in since you are the experts. While we haven't resolved on the initial issue there is a valid proposal for a role of `skeleton` which is a common UX pattern to represent visually that the current content is placeholder content and to expect a certain type of content to appear. This can currently be "worked around" by either denoting it as presentational or giving it a custom label and making the rest of the content presentational. However, this is not always done across libs or by web authors and so it should be considered as an additional role. The discussion in Open UI is here: https://github.com/WICG/open-ui/issues/169

**Comment by carmacleod** [comment]
I wonder if [aria-busy](https://w3c.github.io/aria/#aria-busy) would be helpful?
i.e. `aria-busy="true"` on the thing that isn't ready yet, and change it to `aria-busy="false"` (or simply remove the attribute) when the thing is no longer in a skeleton state.

**Comment by carmacleod** [comment]
(By the way, on a related/unrelated note, if you are writing up recommendations for skeleton state please add a point about supporting the user's prefers-reduced-motion setting so that those moving skeletons don't accidentally make anybody sick). :)

**Comment by una** [comment]
Looking further into this, and acknowledging `aria-busy="true"` and `aria-live="polite"`:

I still feel `role="skeleton"` would still be a useful addition since essentially the element will disappear upon its content being loaded, so the "busy" boolean doesn't *quiet* describe the skeleton component if it's a replaced element, and not on a parent.

**Comment by aardrian** [comment]
@una is it fair to say you are looking at defining a role to account for the pattern here?

https://vuetifyjs.com/en/components/skeleton-loaders/

If so, is it partly because that example has issues (poor DX, little support, conflicts, etc.) or because that pattern has some limitations beyond the replace part?

> essentially the element will disappear upon its content being loaded, so the "busy" boolean doesn't quiet describe the skeleton component if it's a replaced element, and not on a parent.

If the skeleton visual needs to go away, `aria-hidden` can do that work. If it gets removed from the DOM completely, then if the new content is the following sibling it may not be an issue.

**Comment by mathalete** [comment]
Interesting discussion - if it helps at all, I recently crossed swords with similar 'skeleton' elements within a design system. These loading animations had initially been implemented as individual live regions each with the word "Loading" added as the live region announcement.

In this scenario I was looking at, there could be lots of these elements in the document at the same time, so it was a pretty verbose listening experience, it also wasn't clear 'what' exactly was loading, just that the word loading was announced a lot.

We arrived at the following conclusion  -

- not to implement each loading animations as an individual live region - instead just use a global polite region to indicate the view or part of the view was still loading
- add visually hidden text to each loading animation so that if a screen reader user was exploring a specific loading area of the page, it'd be clear that specific contextualised element was in a loading state
- support the users preference for reduced motion, so provide a non-animated equivalent
- ideally ensure the visual design should account for the WCAG 2.1 SC for non-text contrast 1.4.11

**Comment by una** [comment]
@aardrian Yes, that type of component is what I mean here. TO @mathalete's points, it might be easier for implementors to correctly build the experience if there was a singular role they could use (`skeleton`) to define and element which:

1. Is a placeholder for some content that is actively being rendered
2. Will be removed or replaced with said content

**Comment by aardrian** [comment]
@una thanks for that context. Allow me to run through the current situation and see if it fits your need or where it falls short.

* For the content in the process of being rendered, [`aria-busy`](https://www.w3.org/TR/wai-aria-1.1/#aria-busy) conveys that information to assistive technology (screen readers, really) through the accessibility APIs in conforming UAs.
* For placeholder content, there are many methods to provide it — if any is needed depending on the use case — and scripting would need to toggle something like [`aria-hidden`](https://www.w3.org/TR/wai-aria-1.1/#aria-hidden) on that content or otherwise destroy the nodes in the DOM.

A new role on its own may not necessarily make that easier for developers, though it may provide guidance on a pattern to follow (as would an APG pattern or W3C idiom). A new role for users will add verbosity (cues, instructions, etc), but script will still need to go through and remove/replace content regardless of the presence/absence of that role.

I have a prototype that manages those two ARIA states (`aria-busy`, `aria-hidden`) and I think achieves what you want (assuming I understand your objective correctly):
https://codepen.io/aardrian/pen/yLOoOdY

Debug view:
https://cdpn.io/aardrian/debug/yLOoOdY

When you try the debug view in your target screen reader (should work across VO macOS/iOS/iPadOS, TalkBack, JAWS, NVDA, Narrator, Orca and whatever browsers you normally pair with those), does it announce what you expect? If not, what is it missing?

**Comment by scottaohara** [comment]
We discussed this proposal during yesterday's ARIA WG call, and while we appreciate the intent, we do not think there is a need for this role.  

Patterns like this are meant to represent loading content (which could be _any_ type of content as the OPEN UI original issue lists a variety), and there are currently ways to indicate loading states which wouldn't require the creation of a new role.

For instance, the mentioned, `aria-live` / `aria-busy`, visually hidden text - or similarly alt text even if you used images/svgs/similar. This can be done now without the need for a role that people (developers but more importantly users) would have to learn what it means and what sort of expectations they should have for it.

Thanks
